Implement replacements in FormSpec

xrotwang opened the pull request:
Note that this already increases the complexity of what `FormSpec.clean`
does: Now, stripping bracketed content depends on the result of the
replacements. This may be a feature, but could be a trap.

Closes #142

xrotwang commented:
I'm inclined to add multi-space normalization (#146) unconditionally - what do you think?

chrzyki commented:
Awesome! Thanks! The 'trap' warning is absolutely justified, e.g. I've run into this myself already in https://github.com/lexibank/ids/blob/master/lexibank_ids.py. As soon as this is merged, I'll update my datasets to use this replacement variant.

chrzyki commented:
> I'm inclined to add multi-space normalization (#146) unconditionally - what do you think?

Yep! I think that would be good!

LinguList commented:
Yes, this looks good, but what do we do with FirstFormOnlySpec? Should we include that with a true/false flag into the FormSpec, as this would also allow us to loose code, and make it traceable what one does, or do we leave it as such?

xrotwang commented:
Ok, I'll add a first-form-only attribute, and flags to do white-space normalization and unicode normalization. As long as this stays below - say - 10 attributes, the class should still be transparent enough.

LinguList commented:
For the methodology part of a lexibank paper, it will be useful to represent our changes in form of a decision tree, as this is basically what we do here, right? The advantage of flags is that, if we can go on with this, we can see and handle as much data as possible in a uniform manner, which is a great advantage of custom code or customized FormSpec classes.

SimonGreenhill commented:
Yes, it could even be useful to log the Formspec to README.md?

LinguList commented:
Yes, @SimonGreenhill, I think that would be helpful, and reflecting what I tried by then in our google doc, where I wrote down what defaults had been used for splitting. Now with FormSpec, we have this all transparently in one place. 

A table that lists attribute / parameter would probably be all we need.

SimonGreenhill commented:
Just came across this in `robinsonap`:

```csv
blagar-182_mango-1,,blagar,182_mango,"biyát ‘wild mango’, mampleʔ ‘domestic mango’",biyát,b i j a t,,Robinson2012,,
blagar-182_mango-2,,blagar,182_mango,"biyát ‘wild mango’, mampleʔ ‘domestic mango’",mampleʔ,m a m p l e ʔ,,Robinson2012,,
```

the second form, split from `biyát ‘wild mango’, mampleʔ ‘domestic mango’`, becomes ` mample?` (i.e. leading space). Can we also lstrip()?

xrotwang commented:
The output of `FormSpec.as_markdown` looks like this:

## Specification of form manipulation


Specification of the value-to-form processing in Lexibank datasets:

The value-to-form processing is divided into two steps, implemented as methods:
- `FormSpec.split`: Splits a string into individual form chunks.
- `FormSpec.clean`: Normalizes a form chunk.

These methods use the attributes of a `FormSpec` instance to configure their behaviour.

- `brackets`: `{'(': ')'}`
  Pairs of strings that should be recognized as brackets, specified as `dict` mapping opening string to closing string
- `separators`: `;/,`
  Iterable of single character tokens that should be recognized as word separator
- `missing_data`: `('?', '-')`
  Iterable of strings that are used to mark missing data
- `strip_inside_brackets`: `True`
  Flag signaling whether to strip content in brackets (**and** strip leading and trailing whitespace)
- `replacements`: `[]`
  List of pairs (`source`, `target`) used to replace occurrences of `source` in formswith `target` (before stripping content in brackets)
- `first_form_only`: `False`
  Flag signaling whether at most one form should be returned from `split` - effectively ignoring any spelling variants, etc.
- `normalize_whitespace`: `True`
  Flag signaling whether to normalize whitespace - stripping leading and trailing whitespace and collapsing multi-character whitespace to single spaces
- `normalize_unicode`: `None`
  UNICODE normalization form to use for input of `split` (`None`, 'NFD' or 'NFC')

xrotwang commented:
I'd propose to **not** add this to `README.md`, but rather to a saparate `FORMS.md`?

SimonGreenhill commented:
Great idea, thanks @xrotwang. Minor comment, can we change the two iterables  (separators and missing_data) to be the same format by default just for consistency? 

```
separators: ;/,
Iterable of single character tokens that should be recognized as word separator
missing_data: ('?', '-')
```

e.g. separators = (';', '/', ','),

LinguList commented:
@SimonGreenhill, the nature is different though: separators can only be single characters, at least according to the current regex, or do we switch to potential multi-character separators?

SimonGreenhill commented:
ah, ok, if there's an implementation difference, then that's fine. 

(I guess multichar separators are not impossible though: "and", "or")

xrotwang commented:
@SimonGreenhill for `clldutils.split_text_with_content` they are impossible, unfortunately. Do you think this is grave/urgent enough a problem to be addressed right now?

xrotwang commented:
@SimonGreenhill For now, I think, sticking with one-character separators is a lot better defined. For `and`, `or`, etc. there's always the question of requiring enclosing whitespace or not (and possibly other edge cases).

SimonGreenhill commented:
No, not urgently (or ever).

LinguList commented:
I agree with @xrotwang, I am just checking bowernpny, and I found many replacements which are used but occur only, say 10 times in the data. So my rule would be: if we encounter something along those dimensions, we should put more explicit changes into lexemes.tsv to avoid having seemingly targeted code for few exceptions, and this is what we may often encounter (also when people use multi-character separators).

tresoldi commented:
Yes, agreed. And bowernpny was one of the cases of multiple replacements I had in mind, for the obvious reason that I wrote those.

Replacements in FormSpec should be used but not abused, as already discussed.

xrotwang commented:
Ok, I'm merging this now.
